assets/tutorials/2023-07-15-gams-blog.jl

```diff
@@ -0,0 +1,379 @@
+    ijklm[!, :x] = @variable(model, x[1:size(ijklm, 1)] >= 0)
+    for df in DataFrames.groupby(ijklm, :i)
+        @constraint(model, sum(df.x) >= 0)
+    end
+    optimize!(model)
+    return model
+end
+
+@time fast_dataframe_formulation(dataframe_data);
+@time fast_dataframe_formulation(dataframe_data);
+
+# ## Scaling
+
+# Let's now compare the five different formulations over a range of `n` values:
+
+# ```julia
+# import Plots
+#
+# function timings()
+#     N = [100, 200, 400, 700, 1_100, 1_600, 2_200, 2_900]
+#     time_original = Float64[]
+#     time_intuitive = Float64[]
+#     time_fast = Float64[]
+#     time_dataframe = Float64[]
+#     time_fast_dataframe = Float64[]
+#     for n in N
+#         ## Original model
+#         original_data = OriginalData(n)
+#         start = time()
+#         intuitive_formulation(original_data)
+#         push!(time_original, time() - start)
+#         ## Tuple models
+#         tuple_data = TupleData(n)
+#         start = time()
+#         intuitive_formulation(tuple_data)
+#         push!(time_intuitive, time() - start)
+#         start = time()
+#         fast_formulation(tuple_data)
+#         push!(time_fast, time() - start)
+#         ## DataFrame models
+#         dataframe_data = DataFrameData(n)
+#         start = time()
+#         dataframe_formulation(dataframe_data)
+#         push!(time_dataframe, time() - start)
+#         start = time()
+#         fast_dataframe_formulation(dataframe_data)
+#         push!(time_fast_dataframe, time() - start)
+#     end
+#     Plots.plot(
+#         N,
+#         [time_original time_intuitive time_fast time_dataframe time_fast_dataframe];
+#         labels = ["Original" "Intuitive" "Fast" "DataFrame" "Fast DataFrame"],
+#         xlabel = "N",
+#         ylabel = "Solution time (s)",
+#     )
+#     return
+# end
+#
+# timings()
+# ```
+
+# <img src="/assets/tutorials/gams/scaling.svg">
+
+# The dataframe formulations are significantly better. These results demonstrate
+# how benchmarking different modeling systems is difficult, and choosing an
+# appropriate datastructure for a model is too. It's likely that Pyomo and
+# Gurobipy would similarly benefit from using `pandas` to perform the inner join
+# instead of using nested for-loop approach.
+
+# ## Other comments
+
+# There are a few lines in the blog post that deserve some rebuttal.
+
+# > One of the key differences between GAMS and the other modeling frameworks
+# > we’ve mentioned is that GAMS is a domain-specific language
+
+# In our opinion, JuMP is a [domain-specific language](https://en.wikipedia.org/wiki/Domain-specific_language)
+# embedded in a programming language. Indeed, a key feature of JuMP is that the
+# code users write inside the modeling macros (the identifiers beginninng with
+# `@` ) is not what gets executed. Instead, JuMP parses the syntax that that the
+# user writes, and compiles it into a different form that is more efficient.
```

Comment:
I would highlight that this makes use of the metaprogramming ability of Julia so that the reader understand why it's not possible in Python